content/Patterns for maintainable Rails applications. Part II - Ignacio Pérez.md

```diff
@@ -0,0 +1,108 @@
+
+https://gist.github.com/PerezIgnacio/99fa01fd38d910e95dcd0d36ca67ecb9
+
+From Rails 5.1 onwards, however, we can use a helper called delegate_missing_to, which is recommended instead of the above as it is more explicit.
+
+https://gist.github.com/PerezIgnacio/6467f852d165bc569fb7e746c19184a3
+
+https://gist.github.com/PerezIgnacio/0898975690c8b88937536c54ea36eacd
+
+Regardless of how we implement the Decorator, the controller where the model is created should look like the following:
+
+https://gist.github.com/PerezIgnacio/43badd2cf9bdd1210c999ecae89960aa
+
+In case you have been wondering, the Presenter pattern can be considered a sub pattern of the Decorator pattern, as it has the same intent but focuses on trying to keep the logic out of the view. This is implemented similarly to the examples mentioned, but there are also commonly used gems such as Drapper.
+
+### Inheritance/concerns
+
+I included them together because concerns are basically a form of multi-inheritance. Inheritance is useful, but it must be used properly. It must be thought top-down, as creating specialized versions of a parent class and not as a way of extracting duplicated code from two classes. Possible issues with poorly used inheritance are that we may lose explicitness and if we needed to add a third class we would likely need to make unnecessary changes to the hierarchy.
+
+The same applies to concerns, but the consequences are even worse as models can have multiple concerns. Also, in these cases, finding where a method is defined can be tough.
+
+There are also some bad practices related to concerns that should be avoided. An example is extracting methods into a concern just to reduce the size of a class. We may be gaining readability from this but we lose explicitness and responsibilities are still in the class, only that they were hidden in another file. Another example are bidirectional dependencies in concerns, which is the same as assuming that a superclass knows the implementation details of its subclasses, something we should avoid. An even worse version of this last case is dependency between concerns. A proper concern should then be free of these dependencies and have a single well-defined responsibility.
+
+Rails encourages the use of concerns and there are lots of articles that promote them, but its wrong use has led to even be considered an anti-pattern. Quoting Bryan Helmkamp, CEO at CodeClimate: “Any application with an app/concerns directory is concerning.”
+
+Should we really use concerns then? Properly used I don’t think they are that bad, but there are usually more explicit and more scalable alternatives, including any of the patterns mentioned in the article. The choice between these depends on the context they are going to be used, but if it doesn’t seem like any applies, you can go for the Services/Service Objects pattern.
```

Comment:
I'd discourage concerns by default (specially for models) because it makes code less maintainable. Maybe for controllers it makes more sense to have concerns.